dune/gdt/spaces/fv/defaultproduct.hh

```diff
@@ -0,0 +1,168 @@
+// This file is part of the dune-gdt project:
+//   http://users.dune-project.org/projects/dune-gdt
+// Copyright holders: Felix Schindler
+// License: BSD 2-Clause License (http://opensource.org/licenses/BSD-2-Clause)
+//
+// Contributors: Tobias Leibner
+
+#ifndef DUNE_GDT_SPACES_FV_DEFAULTPRODUCT_HH
+#define DUNE_GDT_SPACES_FV_DEFAULTPRODUCT_HH
+
+#include <tuple>
+
+#include <dune/gdt/mapper/default/productfv.hh>
+
+#include <dune/gdt/spaces/productinterface.hh>
+
+#include "default.hh"
+#include "interface.hh"
+
+namespace Dune {
+namespace GDT {
+namespace Spaces {
+namespace FV {
+
+
+// forward, to be used in the traits and to allow for specialization
+template< class GridViewImp, class RangeFieldImp, size_t rangeDim, size_t rangeDimCols = 1 >
+class DefaultProduct
+{
+  static_assert(Dune::AlwaysFalse< GridViewImp >::value, "Untested for these dimensions!");
+};
+
+
+namespace internal {
+
+// from https://stackoverflow.com/questions/16853552/how-to-create-a-type-list-for-variadic-templates-that-contains-n-times-the-sam
+
+// in the end, we would like to have something like indices< 1, 2, 3 > for N = 3
+template< std::size_t... >
+struct indices {};
+
+// we want to call this with empty Indices, i.e. create_indices< N >::type == indices< 1, 2, 3 > for N = 3
+template< std::size_t N, std::size_t... Indices>
+struct create_indices : create_indices< N-1, N-1, Indices...> {};
+
+// terminating template
+template< std::size_t... Indices >
+struct create_indices< 0, Indices...> {
+  typedef indices<Indices...> type;
+};
+
+// T_aliased< T, Index > is always the type T, no matter what Index is
+template<typename T, std::size_t index>
+using T_aliased = T;
+
+// make_identical_tuple< T, N >::type is a std::tuple< T, ... , T > with a length of N
+template< typename T, std::size_t N, typename I = typename create_indices< N >::type >
+struct make_identical_tuple;
+
+template< typename T, std::size_t N, std::size_t ...Indices >
+struct make_identical_tuple< T, N, indices< Indices... > >
+{
+    using type = std::tuple<T_aliased<T, Indices>...>;
+};
+
+
+
+template< class GridViewImp, class RangeFieldImp, size_t rangeDim, size_t rangeDimCols >
+class DefaultProductTraits
+    : public DefaultTraits< GridViewImp, RangeFieldImp, rangeDim, rangeDimCols >
+{
+  typedef DefaultTraits< GridViewImp, RangeFieldImp, rangeDim, rangeDimCols >  BaseType;
+public:
+  typedef DefaultProduct< GridViewImp, RangeFieldImp, rangeDim, rangeDimCols > derived_type;
+  using BaseType::polOrder;
```

Comment:
You only have to pull in the definitions from the `BaseType` by using `using` if you want to use them locally (e.g., in a function in this class of to define another type). Everyone  accessing these types from the outside (e.g., `typename DefaultProductTraits< ... >::GridViewType` will be able to do so without you using `using`, the compiler just takes the definition from the `BaseType`. So you can probably drop a lot of these...